crates/ibc/src/core/ics02_client/error.rs

```diff
@@ -1,280 +1,186 @@
-            { reason: String }
-            | e | {
-                format_args!("Insufficient overlap {}", e.reason)
-            },
-
-        RawClientAndConsensusStateTypesMismatch
-            {
-                state_type: ClientType,
-                consensus_type: ClientType,
-            }
-            | e | {
-                format_args!("mismatch in raw client consensus state {} with expected state {}",
-                    e.state_type, e.consensus_type)
-            },
-
-        LowHeaderHeight
-            {
-                header_height: Height,
-                latest_height: Height
-            }
-            | e | {
-                format!("received header height ({}) is lower than (or equal to) client latest height ({})",
-                    e.header_height, e.latest_height)
-            },
-
-        LowUpgradeHeight
-            {
-                upgraded_height: Height,
-                client_height: Height,
-            }
-            | e | {
-                format_args!("upgraded client height {} must be at greater than current client height {}",
-                    e.upgraded_height, e.client_height)
-            },
-
-        InvalidConsensusStateTimestamp
-            {
-                time1: Timestamp,
-                time2: Timestamp,
-            }
-            | e | {
-                format_args!("timestamp is invalid or missing, timestamp={0},  now={1}", e.time1, e.time2)
-            },
-
-        HeaderNotWithinTrustPeriod
-            {
-                latest_time:Timestamp,
-                update_time: Timestamp,
-            }
-            | e | {
-                format_args!("header not withing trusting period: expires_at={0} now={1}", e.latest_time, e.update_time)
-            },
-
-        MissingLocalConsensusState
-            { height: Height }
-            | e | { format_args!("the local consensus state could not be retrieved for height {}", e.height) },
-
-        InvalidConnectionEnd
-            [ TraceError<TendermintProtoError>]
-            | _ | { "invalid connection end" },
-
-        InvalidChannelEnd
-            [ TraceError<TendermintProtoError>]
-            | _ | { "invalid channel end" },
-
-        InvalidAnyClientState
-            [ TraceError<TendermintProtoError>]
-            | _ | { "invalid any client state" },
-
-        InvalidAnyConsensusState
-            [ TraceError<TendermintProtoError> ]
-            | _ | { "invalid any client consensus state" },
-
-        Signer
-            [ SignerError ]
-            | _ | { "failed to parse signer" },
-
-        Ics23Verification
-            [ Ics23Error ]
-            | _ | { "ics23 verification failure" },
-
-        ClientSpecific
-            { description: String }
-            | e | { format_args!("client specific error: {0}", e.description) },
+#[derive(Debug, Display)]
+pub enum ClientError {
+    /// unknown client type: `{client_type}`
+    UnknownClientType { client_type: String },
```

Comment:
Having thought about this some more, I like the direction this PR is taking. IMO what's left to be done is:
1. Remove all unused errors such as this one. Once this is done, we can additionally look to combine redundant errors.
2. "Roll our own backtrace", as @hu55a1n1 [suggested](https://github.com/cosmos/ibc-rs/pull/255#discussion_r1030775147). We should make sure to use the same convention everywhere, as you started doing here: `<error message>: {0}` (for when the error argument is another `Error`). When `std::error::Error` makes it into `core`, we can start using `thiserror`.
3. Make sure there's a `Error::Other` variant for host methods to use. This PR is getting big enough, we can change host functions in another PR.

Reply:
> When `std::error::Error` makes it into core, we can start using `thiserror`
Currently `thiserror` can not be used in no_std, `thiserror::Error` I think the implementation of this macro can not be used in the current no_std environment, and the current use of `displaydoc` I think similar to `thiserror` except that there is no error backtrace, but here we have implemented `std::error::Error`.

```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match &self {
            Error::InvalidHeader { error: e, .. } => Some(e),
            Error::InvalidTendermintTrustThreshold(e) => Some(e),
            Error::InvalidChainIdentifier(e) => Some(e),
            Error::InvalidChainId { error: e, .. } => Some(e),
            Error::InvalidRawHeader(e) => Some(e),
            Error::Decode(e) => Some(e),
            Error::TimestampOverflow(e) => Some(e),
            _ => None,
        }
    }
}
```

Reply:
when `thiserror` support we just add `#derive[thiserror::Error]` can remove below this.
```rust
#[cfg(feature = "std")]
impl std::error::Error for Error {}
```